sklearn/feature_extraction/tests/test_text.py

```diff
@@ -1042,6 +1045,29 @@ def test_vectorizer_string_object_as_input():
             ValueError, message, vec.transform, "hello world!")
 
 
+@pytest.mark.parametrize("X_dtype", [np.float32, np.float64])
+def test_tfidf_transformer_type(X_dtype):
+    X = rand(10, 20000, dtype=X_dtype, random_state=42)
+    X_trans = TfidfTransformer().fit_transform(X)
+    assert X_trans.dtype == X.dtype
+
+
+@pytest.mark.parametrize(
+    "vectorizer_dtype, output_dtype, expected_warning, msg_warning",
+    [(np.int32, np.float64, UserWarning, "'dtype' should be used."),
+     (np.int64, np.float64, UserWarning, "'dtype' should be used."),
+     (np.float32, np.float32, None, None),
+     (np.float64, np.float64, None, None)]
+)
+def test_tfidf_vectorizer_type(vectorizer_dtype, output_dtype,
+                               expected_warning, msg_warning):
+    X = np.array(["numpy", "scipy", "sklearn"])
+    vectorizer = TfidfVectorizer(dtype=vectorizer_dtype)
+    with pytest.warns(expected_warning, match=msg_warning):
+            X_idf = vectorizer.fit_transform(X)
```

Comment:
To check that no warning is raised (if that's what you want to do) you could use https://stackoverflow.com/a/45671804/1791279. I wonder if,
```py
    with pytest.warns(expected_warning, match=msg_warning) as record:
            X_idf = vectorizer.fit_transform(X)
    if expected_warning is None:
        assert len(record) == 0
```
would work.

Reply:
Right, I was expecting that it works but actually we would need this: https://github.com/pytest-dev/pytest/issues/1830#issuecomment-378357212 to be implemented.